### PR TITLE
Making _unwrap() a method of the ColletionInterface

### LINKED ISSUE #6523: CollectionTrait::append call to _unwrap
This is how `CollectionTrait::append` is implemented now:

    1. public function append($items)
    2. {
    3.     $items = $items instanceof Iterator ? $items : new Collection($items);
    4.     $list = new AppendIterator;
    5.     $list->append($this);
    6.     $list->append($items->_unwrap());
    7.     return new Collection($list);
    8. }

And there are two issues with `$items->_unwrap()` call on line 6:
  1. if `$items` is an instance of `Iterator` (see line 3), it crashes, because Iterator instances usually don't have public `_unwrap` method (only `Collection` instances do),
  2. `CollectionTrait::_unwrap` method is protected, which means it can only be invoked if `$this` and `$items` are instances of same class (I wanted to append something to `\Cake\ORM\ResultSet`, it's not possible to invoke `Collection::_unwrap` from `ResultSet`).

Therefore `append` basically only works on `Collection` instances and only if an array or another `Collection` instance is being appended.

Easily solvable if call to `_unwrap` on line 6 is deleted (see https://github.com/sukihub/cakephp/commit/3a0d548e1a69eecafb788a0abe530a735238cc17). Could there be any unexpected side effects? All tests pass after the change.

### PR DESCRIPTION
**Pull request by lorenzo**
closes #6523

### CODE REVIEW

File: src/Collection/CollectionTrait.php [hunk 1]
```diff
@@ -540,13 +539,12 @@ public function isEmpty()
...
+
     /**
-     * Returns the closest nested iterator that can be safely traversed without
-     * losing any possible transformations.
+     * {@inheritDoc}
      *
-     * @return \Iterator
      */
-    protected function _unwrap()
+    public function _unwrap()
```
Comment (dereuromark) [inline src/Collection/CollectionTrait.php]: Shouldnt we keep this method and rather make a new wrapper `public function unwrap()` around it? Would be less akward looking.

File: src/Collection/CollectionTrait.php [hunk 2]
```diff
@@ -316,10 +316,9 @@ public function first()
      */
     public function append($items)
     {
-        $items = $items instanceof Iterator ? $items : new Collection($items);
         $list = new AppendIterator;
         $list->append($this);
```
Comment (sukihub) [inline src/Collection/CollectionTrait.php]: Wouldn't `$this->unwrap()` be a good idea?

**Comment by markstory** [comment]
Looks good :+1:

**Comment by sukihub** [comment]
I also have some tests, if you are interested https://github.com/sukihub/cakephp/commit/3a0d548e1a69eecafb788a0abe530a735238cc17

**Comment by markstory** [comment]
@sukihub Send the tests as a pull request :)

File: src/Collection/CollectionTrait.php (hunk 2 above)
Comment (lorenzo) [inline src/Collection/CollectionTrait.php]: sure, send it as a pull request if you like. I can do it on my evening otherwise :)

**Pull request merged**